samza-core/src/main/java/org/apache/samza/operators/impl/PartialJoinOperatorImpl.java

```diff
@@ -20,28 +20,70 @@
 
 import org.apache.samza.config.Config;
 import org.apache.samza.operators.MessageStreamImpl;
+import org.apache.samza.operators.functions.PartialJoinFunction;
+import org.apache.samza.operators.functions.PartialJoinFunction.PartialJoinMessage;
 import org.apache.samza.operators.spec.PartialJoinOperatorSpec;
+import org.apache.samza.storage.kv.Entry;
+import org.apache.samza.storage.kv.KeyValueIterator;
+import org.apache.samza.storage.kv.KeyValueStore;
 import org.apache.samza.task.MessageCollector;
 import org.apache.samza.task.TaskContext;
 import org.apache.samza.task.TaskCoordinator;
 
+import java.util.ArrayList;
+import java.util.List;
 
 /**
- * Implementation of a {@link PartialJoinOperatorSpec}. This class implements function
- * that only takes in one input stream among all inputs to the join and generate the join output.
+ * Implementation of a {@link PartialJoinOperatorSpec} that joins messages of type {@code M} in this stream
+ * with buffered messages of type {@code JM} in the other stream.
  *
  * @param <M>  type of messages in the input stream
  * @param <JM>  type of messages in the stream to join with
  * @param <RM>  type of messages in the joined stream
  */
-class PartialJoinOperatorImpl<M, K, JM, RM> extends OperatorImpl<M, RM> {
+class PartialJoinOperatorImpl<K, M, JM, RM> extends OperatorImpl<M, RM> {
 
-  PartialJoinOperatorImpl(PartialJoinOperatorSpec<M, K, JM, RM> joinOp, MessageStreamImpl<M> source, Config config, TaskContext context) {
-    // TODO: implement PartialJoinOperatorImpl constructor
+  private final PartialJoinFunction<K, M, JM, RM> thisPartialJoinFn;
+  private final PartialJoinFunction<K, JM, M, RM> otherPartialJoinFn;
+  private final long ttlMs;
+
+  PartialJoinOperatorImpl(PartialJoinOperatorSpec<K, M, JM, RM> partialJoinOperatorSpec, MessageStreamImpl<M> source,
+      Config config, TaskContext context) {
+    this.thisPartialJoinFn = partialJoinOperatorSpec.getThisPartialJoinFn();
+    this.otherPartialJoinFn = partialJoinOperatorSpec.getOtherPartialJoinFn();
+    this.ttlMs = partialJoinOperatorSpec.getTtlMs();
   }
 
   @Override
   public void onNext(M message, MessageCollector collector, TaskCoordinator coordinator) {
-    // TODO: implement PartialJoinOperatorImpl processing logic
+    K key = thisPartialJoinFn.getKey(message);
+    thisPartialJoinFn.getState().put(key, new PartialJoinMessage<>(message, System.currentTimeMillis()));
+    PartialJoinMessage<JM> otherMessage = otherPartialJoinFn.getState().get(key);
+    if (otherMessage != null) {
+      RM joinResult = thisPartialJoinFn.apply(message, otherMessage.getMessage());
+      this.propagateResult(joinResult, collector, coordinator);
+    }
+  }
+
+  @Override
+  public void onTimer(MessageCollector collector, TaskCoordinator coordinator) {
+    KeyValueStore<K, PartialJoinMessage<M>> thisState = thisPartialJoinFn.getState();
+    KeyValueIterator<K, PartialJoinMessage<M>> iterator = thisState.all();
```

Comment:
@prateekm sure, we can postpone this change. But I would strongly prefer to have some metrics/indicator to show the time spent in onTimer(). Main concern here is that since the state is fully controlled by us, user won't see it and when the container gets really slow, we need a fast way to detect the issue if the latency is too long in onTimer(). However, per operator metrics is also a big scope of work. I think that I am OK for now to just add two log lines at the beginning and ending of the onTimer() to report the time latency. We definitely need to design the per operator metrics soon.